src/value/loader.rs

```diff
@@ -0,0 +1,462 @@
+// Copyright Amazon.com, Inc. or its affiliates.
+
+use crate::result::IonResult;
+use crate::value::owned::{OwnedElement, OwnedSequence, OwnedStruct, OwnedSymbolToken, OwnedValue};
+use crate::value::AnyInt;
+use crate::IonType;
+use ion_c_sys::reader::{IonCReader, IonCReaderHandle};
+use ion_c_sys::ION_TYPE;
+use std::convert::{TryFrom, TryInto};
+
+use super::owned::text_token;
+
+/// TODO add/refactor trait/implementation for borrowing over some context
+///      we could make it generic with generic associated types or just have a lifetime
+///      scoped implementation
+
+/// Loads Ion data into [`Element`] instances.
+///
+/// Users of this trait should not assume any particular implementation of `Element`.
+/// In the future, [generic associated types (GAT)][gat] and [existential types in traits][et]
+/// should make it easier to model this more abstractly.
+///
+/// [gat]: https://rust-lang.github.io/rfcs/1598-generic_associated_types.html
+/// [et]:https://rust-lang.github.io/rfcs/2071-impl-trait-existential-types.html
+pub trait Loader {
+    /// Parses Ion over a given slice of data and yields each top-level value as
+    /// an [`Element`] instance.
+    ///
+    /// The [`Iterator`] will generally return `Ok([Element])` but on a failure of
+    /// parsing it will return an `Err([IonError])` and then a `None` to signal no more
+    /// elements.
```

Comment:
This makes sense conversationally, but I suggest preserving the `Option` wrapper when discussing the return values.

```suggestion
    /// The [`Iterator`] will generally return `Some(Ok([Element]))` but on a failure of
    /// parsing it will return a `Some(Err([IonError]))` and then a `None` to signal no more
    /// elements.
```